hail/src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -416,38 +416,47 @@ object Simplify {
     case InsertFields(SelectFields(old, _), Seq(), Some(insertFieldOrder)) =>
       SelectFields(old, insertFieldOrder)
 
-    case Let(Seq(), body) =>
+    case Block(Seq(), body) =>
       body
 
-    case Let(xs, Let(ys, body)) =>
-      Let(xs ++ ys, body)
+    case Block(xs, Block(ys, body)) =>
+      Block(xs ++ ys, body)
 
     // assumes `NormalizeNames` has been run before this.
-    case Let(Let.Nested(before, after), body) =>
-      def numBindings(b: (String, IR)): Int =
-        b._2 match {
-          case let: Let => 1 + let.bindings.length
+    case Block(bindings, body) if bindings.exists(_.value.isInstanceOf[Block]) =>
+      def numBindings(b: Binding): Int =
+        b.value match {
+          case let: Block => 1 + let.bindings.length
           case _ => 1
         }
 
       val newBindings =
-        new BoxedArrayBuilder[(String, IR)](
-          after.foldLeft(before.length)((sum, binding) => sum + numBindings(binding))
+        new BoxedArrayBuilder[Binding](
+          bindings.foldLeft(0)((sum, binding) => sum + numBindings(binding))
         )
 
-      newBindings ++= before
-
-      after.foreach {
-        case (name: String, ir: Let) =>
-          newBindings ++= ir.bindings
-          newBindings += name -> ir.body
-        case (name, value) =>
-          newBindings += name -> value
+      bindings.foreach {
+        case Binding(name, ir: Block, scope) =>
+          newBindings ++= (if (scope == Scope.EVAL) ir.bindings
+                           else ir.bindings.map {
+                             case Binding(name, value, Scope.EVAL) => Binding(name, value, scope)
+                             case _ => fatal("Simplify: found nested Agg bindings")
```

Comment:
Futhermore, we shouldn't leave this sort of thing to the simplifier

Reply:
This is just not possible under our scoping rules. The default case is dead code on any input that typechecks.

The reason this isn't possible is because in the value of an agg binding, the agg environment is promoted to eval, leaving a `None` agg environment, disallowing agg bindings.